File: frontend/AutoTracker/Managers/Validator.swift
import Foundation

class Validator {
    static func isValidEmail(for email: String) -> Bool {
        let email = email.trimmingCharacters(in: .whitespacesAndNewlines)
        let emailRegEx = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.{1}[A-Za-z]{2,64}"
        let emailPred = NSPredicate(format: "SELF MATCHES %@", emailRegEx)
        return emailPred.evaluate(with: email)
    }
    
    static func isValidDate(for birthday: String) -> Bool {
        let birthday = birthday.trimmingCharacters(in: .whitespacesAndNewlines)
        let birthdayRegEx = "[0-9]{2}\\.[0-9]{2}\\.[0-9]{4}"
        let birthdayPred = NSPredicate(format: "SELF MATCHES %@", birthdayRegEx)
        return birthdayPred.evaluate(with: birthday)
    }
    
    static let minPswdLen = 8
    static func isValidPswd(for pswd: String) -> Bool {
        let pswd = pswd.trimmingCharacters(in: .whitespacesAndNewlines)
        let pswdRegEx = ".{\(minPswdLen),}"
        let pswdPred = NSPredicate(format: "SELF MATCHES %@", pswdRegEx)
        return pswdPred.evaluate(with: pswd)
    }
    
    static func isValidGosNumMask(for gosNumMask: String) -> Bool {
        let gosNumMask = gosNumMask.trimmingCharacters(in: .whitespacesAndNewlines)
        let gosNumMaskRegEx = "^([АВЕКМНОРСТУХ*])(\\d|\\*){3}([АВЕКМНОРСТУХ*]{2})(\\d{2,3}|\\*{1})$"
        let gosNumMaskPred = NSPredicate(format: "SELF MATCHES %@", gosNumMaskRegEx)
        return gosNumMaskPred.evaluate(with: gosNumMask)
    }
    
    static func isValidPassport(for passport: Document) -> Bool {
        passport.number.count == 6 && passport.serial.count == 4
    }
    
    static func isValidGosNum(for gosNum: String) -> Bool {
        let gosNum = gosNum.trimmingCharacters(in: .whitespacesAndNewlines)
        let gosNumRegEx = "^[АВЕКМНОРСТУХ]\\d{3}[АВЕКМНОРСТУХ]{2}\\d{2,3}$"
        let gosNumPred = NSPredicate(format: "SELF MATCHES %@", gosNumRegEx)
        return gosNumPred.evaluate(with: gosNum)
    }
}

File: backend/api/src/business_logic/src/validator.rs
use models::Document;
use regex::Regex;

pub struct Validator;

impl Validator {
    pub fn is_valid_email(email: &String) -> bool {
        let re = Regex::new(r"^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$").unwrap();
        re.is_match(email)
    }

    pub fn is_valid_password(pswd: &String) -> bool {
        let re = Regex::new(r".{8,}").unwrap();
        re.is_match(pswd)
    }

    fn is_number(s: &str) -> bool {
        s.chars().all(|c| c.is_ascii_digit())
    }

    pub fn is_valid_passport(passport: &Document) -> bool {
        let mut is_correct = false;
        if passport.serial.len() == 4 && passport.number.len() == 6 {
            if Self::is_number(&passport.serial) && Self::is_number(&passport.number) {
                is_correct = true;
            }
        }
        is_correct
    }

    pub fn is_valid_gos_num(gos_num: &String) -> bool {
        let re = Regex::new(r"^[АВЕКМНОРСТУХ]\d{3}[АВЕКМНОРСТУХ]{2}\d{2,3}$").unwrap();
        re.is_match(gos_num)
    }

    pub fn is_valid_gos_num_mask(gos_num_mask: &String) -> bool {
        let re = Regex::new(r"^([АВЕКМНОРСТУХ*])(\d|\*){3}([АВЕКМНОРСТУХ*]{2})(\d{2,3}|\*{1})$")
            .unwrap();
        re.is_match(gos_num_mask)
    }

    pub fn is_valid_date(date: &String) -> bool {
        let re = Regex::new(r"^\d{2}\.\d{2}\.\d{4}$").unwrap();
        re.is_match(date)
    }
    pub fn is_valid_time(time: &String) -> bool {
        let re = Regex::new(r"^\d{1,2}\:\d{2}$").unwrap();
        if !re.is_match(time) {
            return false;
        }

        let parts: Vec<&str> = time.split(':').collect();
        let hours: u32 = parts[0].parse().unwrap_or(24);
        let minutes: u32 = parts[1].parse().unwrap_or(60);

        hours < 24 && minutes < 60
    }
}
